pkg/nodeagent/controller/operatingsystemconfig/reconciler.go

```diff
@@ -0,0 +1,327 @@
+func (r *Reconciler) applyChangedFiles(log logr.Logger, files []extensionsv1alpha1.File) error {
+	tmpDir, err := r.FS.TempDir("", "gardener-node-agent-*")
+	if err != nil {
+		return fmt.Errorf("unable to create temporary directory: %w", err)
+	}
+	defer func() { utilruntime.HandleError(r.FS.RemoveAll(tmpDir)) }()
+
+	for _, file := range files {
+		if file.Content.Inline == nil {
+			continue
+		}
+
+		if err := r.FS.MkdirAll(filepath.Dir(file.Path), fs.ModeDir); err != nil {
+			return fmt.Errorf("unable to create directory %q: %w", file.Path, err)
+		}
+
+		permissions := defaultFilePermissions
+		if file.Permissions != nil {
+			permissions = fs.FileMode(*file.Permissions)
+		}
+
+		data, err := extensionsv1alpha1helper.Decode(file.Content.Inline.Encoding, []byte(file.Content.Inline.Data))
+		if err != nil {
+			return fmt.Errorf("unable to decode data of file %q: %w", file.Path, err)
+		}
+
+		tmpFilePath := filepath.Join(tmpDir, filepath.Base(file.Path))
+		if err := r.FS.WriteFile(tmpFilePath, data, permissions); err != nil {
+			return fmt.Errorf("unable to create temporary file %q: %w", tmpFilePath, err)
+		}
+
+		if err := r.FS.Rename(tmpFilePath, file.Path); err != nil {
+			return fmt.Errorf("unable to rename temporary file %q to %q: %w", tmpFilePath, file.Path, err)
+		}
+
+		log.Info("Successfully applied new or changed file", "path", file.Path)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) removeDeletedFiles(log logr.Logger, files []extensionsv1alpha1.File) error {
+	for _, file := range files {
+		if err := r.FS.Remove(file.Path); err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+			return fmt.Errorf("unable to delete no longer needed file %q: %w", file.Path, err)
+		}
+
+		log.Info("Successfully removed no longer needed file", "path", file.Path)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) applyChangedUnits(ctx context.Context, log logr.Logger, units []changedUnit) error {
+	for _, unit := range units {
+		unitFilePath := path.Join(etcdSystemdSystem, unit.Name)
+
+		if unit.Content != nil {
+			oldUnitContent, err := r.FS.ReadFile(unitFilePath)
+			if err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+				return fmt.Errorf("unable to read existing unit file %q for %q: %w", unitFilePath, unit.Name, err)
+			}
+
+			newUnitContent := []byte(*unit.Content)
+			if !bytes.Equal(newUnitContent, oldUnitContent) {
+				if err := r.FS.WriteFile(unitFilePath, newUnitContent, defaultFilePermissions); err != nil {
+					return fmt.Errorf("unable to write unit file %q for %q: %w", unitFilePath, unit.Name, err)
+				}
+				log.Info("Successfully applied new or changed unit file", "path", unitFilePath)
+			}
+
+			// ensure file permissions are restored in case somebody changed them manually
+			if err := r.FS.Chmod(unitFilePath, defaultFilePermissions); err != nil {
+				return fmt.Errorf("unable to ensure permissions for unit file %q for %q: %w", unitFilePath, unit.Name, err)
+			}
+		}
+
+		for _, dropIn := range unit.DropIns {
```

Comment:
```suggestion
		for _, dropIn := range unit.dropIns.changed {
```
Shouldn't this loop over the changed dropins?
Also, a handling for `unit.dropIns.deleted` is missing.